### ISSUE TITLE
Overriding baseUrl from cypress.env.json

### ISSUE THREAD

**Issue by Vages** [cypress-io/cypress#909]
- Operating System: Mac OS 10.13.1 (17B48)
- Cypress Version: Cypress package version: 1.0.3
- Browser Version: Chrome 60

### Is this a Feature or Bug?
Bug

### Current behavior:
Setting `baseUrl` in `cypress.env.json` leads to the key `baseUrl` being created inside `env` (the environment variables), not in the existing `baseUrl` being overridden at the configuration variable level.

### Desired behavior:
`baseUrl` should be overridden, the behavior matching that stated in the [section "Overriding Configuration" in the environment variables documentation](https://docs.cypress.io/guides/guides/environment-variables.html#Overriding-Configuration). (Note: Overriding via the command line works as expected.)

### How to reproduce/Test Code:

cypress.json:
```json
{
  "baseUrl": "http://example.com/setFromCypress.json",
  "env": {
    "someVariable": "originalValue"
  }
}
```

cypress.env.json:
```json
{
  "baseUrl": "http://example.com/setFromCypress.env.json",
  "someVariable": "newValue"
}
```
Screenshot of the resulting configuration:
<img width="949" alt="screen shot 2017-11-13 at 11 28 44" src="https://user-images.githubusercontent.com/3520744/32780576-143d4ad8-c942-11e7-80b9-ec1c783470d8.png">

### Additional Info (images, stack traces, etc)
This issue was initially posted and phrased as a question on StackOverflow: https://stackoverflow.com/q/47262338/1858138

**Comment by brian-mann** [comment]
This isn't a bug. `cypress.env.json` is strictly for environment variables - not configuration.

https://docs.cypress.io/guides/guides/environment-variables.html#Setting-Environment-Variables

To override configuration you'll need to override it from the command line or by using special environment variables prefixed with `CYPRESS_<config var>`

**Comment by dwelle** [comment]
From [docs](https://docs.cypress.io/guides/guides/environment-variables.html#Overriding-Configuration), I got the same idea as OP. This part in particular:

> If your environment variables match a standard configuration key, then instead of setting an environment variable they will instead override the configuration value.

Plus, the cypress runner configuartion "settings" tab doesn't do a good job in explaining that even though `env` has higher precedence than e.g. `cypress.json` or `default`, it doesn't override it at all...

**Comment by Vages** [comment]
But if this is set in stone, I can update the docs to express this more clearly. Just say the word, @brian-mann.

**Comment by DanAtShenTech** [comment]
@brian-mann I'm writing this in order to learn, and I'm respectfully requesting that this issue be re-examined.  From [the docs](https://docs.cypress.io/guides/guides/environment-variables.html#Option-2-cypress-env-json):  
Option #2: cypress.env.json
You can create your own cypress.env.json file that Cypress will automatically check. **Values in here will overwrite conflicting values in cypress.json.**

This strategy is useful because if you add cypress.env.json to your .gitignore file, the values in here can be different for each developer machine.

// cypress.env.json

{
  "host": "veronica.dev.local",
  "api_server": "http://localhost:8888/api/v1/"
}

This documentation states that values in cypress.env.json will overwrite conflicting values in cypress.json, and her[e](https://gitter.im/cypress-io/cypress/archives/2016/11/14) one of your team members @jennifer-shehane writes:

> So, in the Desktop app, when you click into a project, there is a 'Config' tab. That will show you what the baseUrl has been set to and where it is being set from (whether cypress.json, cypress.env.json, set from environment variables or set as CLI arguments) - . . .

I develop on a Vagrant box, and in my current project I have a base url of 127.0.0.1:8090 (with the Vagrant configuration file mapping port 8090 to port 80).  It seems that baseUrl should be able to be set from the cypress.env.json file.  For now, then, I'm doing the following:
`        cyEnv = Cypress.env()`
`        Cypress.config("baseUrl", cyEnv.baseUrl)`

Am I engaging in an anti-pattern here?  I would be glad to propose updated documentation if you would prefer not to allow baseUrl to be overridden in cypress.env.json.

**Comment by brian-mann** [comment]
The documentation could be more specific.

Conflicting **environment variables** will be overriden.

It doesn't make sense to add `baseUrl` to `cypress.env.json` because its not an environment variable, its a configuration value. `cypress.json` and `cypress.env.json` are two different things with different use cases.

If we made `cypress.env.json` override configuration we might as well call it `cypress2.json. As it stands, `cypress.json` needs a nested `env` key to specify environment variables, whereas `cypress.env.json` requires a flat structure where everything becomes an environment variable. These structures are incompatible.

Honestly `cypress.env.json` was not a good idea and it's worth investigating just to remove this entirely. It sounds like people need more complex configuration + env variable needs than what is currently implemented.

Likely we can just add a new plugin event called `config` (we've talked about this internally) and then you can have your own logic to rewrite Cypress' configuration values and environment variables based on whatever logic you want to write in JS, including support async reads. This way you have more programmatic control of whatever it is you want to change.

**Comment by Vages** [comment]
A file which allows any team member to locally override configuration variables and whose internal structure is the same as that of `cypress.json` would fulfill all of my team's needs at this point. By the internal structure being the same, I mean that configuration variables are specified at the first level of the tree structure, and environment variables are defined inside of an object at the key `env`.

By the way, I am very glad that we are having this discussion. Even if I don't get my way here, re-opening the discussion makes me feel like Cypress' creators are genuinely interested in hearing their users' opinions.

**Comment by DanAtShenTech** [comment]
@brian-mann I think that the concept of cypress.env.json was a good one but that perhaps expanding it would be more than adequate at this point (as opposed to creating a new plugin event right now).  Changing the names/function of the files as follows might be a good approach.

- Change cypress.json to cypress.conf.json

- Have a cypress.conf.**local**.json file whose key structure can match and override cypress.conf.json as @Vages suggests and which would not be checked into source control

This structure would allow you to keep the cypress.env.json for backward compatibility at this point, but it could be deprecated and then eventually eliminated (same for cypress.json).

**Comment by brian-mann** [comment]
There have been several other issues opened where people want multiple config files based on their environment - QA, dev, staging, prod. I need to read those to come up with a solution that covers all of these use cases.

I think what we're going to do is deprecate `cypress.env.json` and then give you a new command line option called: `configFile`.

This is what you will use to specify a **different** config file other than `cypress.json`.

However I would make the logic that this other file overrides conflicting values found in `cypress.json` but that `cypress.json` is still read in as the defaults.

Adding a programmatic event for `config` is also going to go in - which gives you even more flexibility to control things using regular JS code as opposed to files and our own arbitrary logic.

**Comment by DanAtShenTech** [comment]
@brian-mann 
> However I would make the logic that this other file overrides conflicting values found in cypress.json but that cypress.json is still read in as the defaults.

Thank you - that would be great.  A question then:  If an overriding config file, as described above, can only be invoked through a command line option, then could it be easily invoked in the local test runner?  (Still learning here.)  Alternatively (and as I mentioned in my last comment), if cypress.conf.json was the default config file and you then looked for any other file of the form cypress.conf.xxxxxxx.json, then the ".xxxxxxxx" could be ".local", ".dev", ".staging", ".prod", or whatever.  Thanks for  your consideration and this outstanding product.

**Comment by bahmutov** [comment]
If I may offer some advice. Overriding configuration settings via environment variables or config files seems like a separate question from running Cypress.

Can we all agree that every configuration setting can be set via an environment variable? If not, open  an issue for that setting.

So the real question is then how to conveniently start cypress with particular (maybe large) set of environment variables to test some specific `baseUrl` and `timeout` etc. For this you can use 3rd party tools like my own [as-a](https://github.com/bahmutov/as-a) - this is the tool we use internally to quickly run any tool with many many custom environment variables.

Just create a file `~/.as-a/.as-a.ini` and put different blocks of settings there. For example, if you want to have a custom `baseUrl`  pointing at staging server and viewport resolution 600 by 700 pixels (see https://docs.cypress.io/guides/references/configuration.html#Environment-Variables)

```
[staging-tablet]
CYPRESS_BASE_URL=https://my-staging.domain.com
CYPRESS_VIEWPORT_WIDTH=600
CYPRESS_VIEWPORT_HEIGHT=700
```
Then run Cypress with command `as-a staging-tablet node_modules/.bin/cypress run`

No need to create hierarchy of configuration JSON files.

**Comment by jennifer-shehane** [comment]
I created a new issue in our [docs](https://github.com/cypress-io/cypress-documentation) to document being clearer about **environment variables** being overridden here: https://github.com/cypress-io/cypress-documentation/issues/236. Our documentation is open source and contributions are welcome. 😄

**Comment by DanAtShenTech** [comment]
Thank you @jennifer-shehane.  I think I've been confused between [baseUrl](https://docs.cypress.io/api/commands/visit.html#Arguments) and the environment variable [host](https://docs.cypress.io/guides/guides/environment-variables.html#).  I realize that baseUrl could be made anything (and in the documentation it actually includes the host), but perhaps it would be more clear if cy.visit() worked like this:
cy.visit(host + baseUrl + url).

**Comment by brian-mann** [comment]
So it sounds like this issue has diverged into 3 distinct issues. I'd like to get buy in that I have this correct and if so I will open these three issues:

1. We will remove `cypress.env.json` entirely as a new breaking version.
2. We will expose a new `config` event in the new Plugin API which has landed as of `1.1.0`. This will be used to programmatically alter any of the configuration or environment variables. Essentially this will be the "last step" before these values are sent to the Driver for use in your tests. This would enable you to do virtually **anything** you want in any kind of capacity. Read in files from your file system, merge, deep merge, overrwrite, hell make an http request to some server for the values, do whatever it is you want to do.
3. Add a new `configFile` CLI option which enables you to point to another arbitrary `cypress.json` file. The use would be `cypress run --config-file ./cypress.staging.json`. These values would override the ones found in `cypress.json`. This enables you to have per machine or per environment files you can control from the CLI or from the Module API.

@DanAtShenTech as per your latest comment regarding `baseUrl` and `host`. I'm not yet seeing the use case or value of this. I guess I understand that you are using `host` as we imagined `baseUrl` would be used. However it seems like the easiest thing for you to do is just to simply use `Cypress.Commands.override` for both `cy.request` and `cy.visit` and then manage this logic yourself. You could just require node's URL module to resolve the arguments passed in and then massage the baseUrl and the host yourself.

If I'm off on this, it would be helpful to have a more concrete example. Perhaps write the code you wish you had (it can be pseudo code). Or maybe even paste some real Cypress code.

**Comment by DanAtShenTech** [comment]
Thank you @brian-mann and @jennifer-shehane.  At this point I don't have more to offer (as far as code or pseudo code) than what I mentioned above about how Django handles things.  My purpose in making the suggestion was not to say that things should simply be done as they are in Django because that is the environment that I work in, but rather, because the Django approach simply seems the most logical way to handle things since every request goes to some host.  However, if Node has a BaseURL concept that is not the same as Django's (in which the base url does not include the host but, rather, comes after the host), then of course a JS testing framework has to lean toward implementing things in a way that makes sense to Node users.  That said, the Django ecosystem is quite large, and I'd like to see other Django developers using Cypress.  I do think that how Django separates host from base url makes the most sense from a logical standpoint, and so that is why I think it would be a good way for Cypress to handle things in the proposed manner as well.

**Comment by DanAtShenTech** [comment]
I just noticed the interchangeable use of "host" and "baseUrl" [in point number 2 here](https://docs.cypress.io/api/commands/request.html#Arguments).  "baseUrl" seems, therefore, to currently be redundant with "host".  Perhaps "baseUrl" could simply be eliminated if you would not like to use it in the above-mentioned fashion.

**Comment by brian-mann** [comment]
We took the concept of `baseUrl` from other JS e2e testing tools (like protractor) - they all work the same way.

`baseUrl` means "anything you want prefixed before your own values".

It could mean: `http://localhost:1234"

It could also mean: "http://localhost:1234/foo/bar/baz"

Whatever it is, we just toss it before what you pass to us in `cy.visit` and `cy.request`.

What I was saying is that you can construct and normalize this behavior yourself.

If you wanted to split out the path from host - you could make baseUrl just the host, and then set a new env var for path.

Upon doing that you could then do something like:

```js
// cypress/support/index.js

const url = require('url')
const baseUrl = Cypress.config('baseUrl')
const host = Cypress.env("HOST")

// reset baseUrl by combining the two
Cypress.config("baseUrl", url.resolve(baseUrl, host))
```

**Comment by tzookb** [comment]
Read the whole discussion and surely seems like a workaround we all need

How would you suggest for having the baseUrl different for each env:

John machine:
john.example.com

Dana machine:
dana.example.com

QA machine:
qa.example.com

I can manipulate it with tricks with env file,
but it doesnt look good.

any other suggestions??

Thanks!

**Comment by bahmutov** [comment]
Easiest for you is to use environment variable when running cypress CYPRESS_baseUrl

**Comment by brian-mann** [comment]
@tzookb yah, using CLI args or our module API would override baseUrl for you based on the environment you specify. That's exactly what the use case of `baseUrl` is. So not seeing how our current implementation isn't working for you.

Regardless this is about to land and you'll have the ability to programmatically alter configuration and environment variables in code. https://github.com/cypress-io/cypress/issues/1042

**Comment by bahmutov** [comment]
@tzookb here is how we do it - pointing same tests at different baseUrl locally, on CI, on staging and after deploying to production: https://www.cypress.io/blog/2017/05/30/cypress-and-immutable-deploys/

**Comment by brian-mann** [comment]
Released in [`1.2.0`](https://docs.cypress.io/guides/references/changelog.html#1-2-0).

**Comment by Konstruktour** [comment]
brian-mann commented on 22 Nov 2017
> Add a new configFile CLI option which enables you to point to another arbitrary cypress.json file. The use would be cypress run --config-file ./cypress.staging.json. These values would override the ones found in cypress.json. This enables you to have per machine or per environment files you can control from the CLI or from the Module API.

This isn't implemented in 1.4.1? Because using it results in an error 'unknown option' :(

**Comment by brian-mann** [comment]
@Konstruktour no but you can do this yourself with the Plugin Configuration API.

https://docs.cypress.io/api/plugins/configuration-api.html#

We'll likely do this at some point - but I'd like to make it part of breaking changes because currently there are too many ways to set things and its overwhelming for users.

**Comment by facetcounter** [comment]
Just for the record, I tried setting CYPRESS_baseUrl in cypress.env.json before I found this thread; that seemed intuitive with the idea of having cypress.env.json override environmental variables set in other ways and with the idea that environmental variables could override config values...

I like the solution of just having a local equivalent to cypress.json and pointing to it using a command line option, and I understand that this solves the same issue in a different way.

The reason I'm mentioning this is that I ended up with a set of environment variables in my runtime that didn't reflect my current configuration... in other words I could see a CYPRESS_baseUrl env var, but the value set in my config file was not overridden. It seems mildly valuable to have this tie be unambiguous if possible.

**Comment by brian-mann** [comment]
@facetcounter I understand what you're going for - it makes sense in theory... but doing it this way is really just creating another `cypress.overrides.json` which was/is not the point of `cypress.env.json`.

We are determined to remove support for `cypress.env.json` in a breaking version down the road and we are instead going to support a `--config-file` argument.

The best programmatic option for you today is the Plugin Configuration API. https://docs.cypress.io/api/plugins/configuration-api.html#

You can literally do anything you want with both config and env vars programmatically under your control in any fashion you desire. That is now the best solution.

**Comment by kud** [comment]
Glad to see I ain't the only one to have a problem with this. (see my issue above)

To be fair, the best way could/should be to accept a javascript file retuning json instead of a json file. Like that we could use for instance `process.env` or `localEnv = require('cypress.local.js)` for baseUrl or whatever. :)

**Comment by westhomas** [comment]
We use docker `.env` files everywhere in our workflow. So I decided to keep that pattern with our cypress tests. Here's my solution:

```javascript
// plugins/index.js

const fs = require('fs')
const path = require('path')

module.exports = (on, config) => {
    // Parse the .env file and inject the settings into the Cypress.env object
    const r1 = fs.readFileSync(path.resolve('.env'), 'utf-8')
    r1.split('\n').map(i => {
        if (i.trim() !== '') {
            const [key, value] = i.split('=')
            config.env[key] = value
        }
    })
    return config
}
```

**Comment by jaypan** [comment]
It doesn't really make sense to me that the testing URL cannot be set per environment in the environment settings. I want to commit cypress.json to GIT, and have cypress.env.json handle environment specific variables. The URL of the testing server is dependent upon the environment. If I put that URL into cypress.json, and commit it to Git, then it will only work on environments that happen to use that same URL. This doesn't make sense to me.

I ended up hacking together a solution based on westhomas' post above. With the below code, if you've set baseUrl in cypress.env.json, it will override whatever you may have set in cypress.json:

```
const fs = require('fs');
const path = require('path');

module.exports = (on, config) => {
  "use strict";

  const file = fs.readFileSync(path.resolve('cypress.env.json'), 'utf-8');
  file.split('\n').map((line) => {
    if (line.trim() !== '') {
      const matches = line.match(/"baseUrl": "(.+)?"/);
      if (matches) {
        config.baseUrl = matches[1];
      }
    }
  });

  return config;
};
```

**Comment by westhomas** [comment]
@jaypan Nice. I was a bit surprised to find out there was no ability to overwrite settings in a cascading manner out-of-the-box. I think that would be helpful.

**Comment by pietmichal** [comment]
Any ETA on making this less confusing?

**Comment by AndreKR** [comment]
Do I understand this correctly? Setting the baseUrl locally should be done using a plugin? Is there some kind of built-in plugin that can read it from a file, maybe even from `cypress.env.json`?

**Comment by dwelle** [comment]
@AndreKR if you don't need to do any overriding, just set it once in `cypress.json`. If you do, then do it in `cypress/plugins/index.js`, see https://docs.cypress.io/api/plugins/configuration-api.html#Usage.

**Comment by AndreKR** [comment]
But `cypress.json` also contains global settings, like the screen size, that I commit to the repository, while the URL is a local setting. What do I put in `.gitignore`?

**Comment by PawelWesolowski** [comment]
It is a terrible idea to remove support for .env.json simply because people are using this strategy to setup local runs. Look at popularity of https://www.npmjs.com/package/dotenv - people like to create a local .dev file and gitignore it.
Problem with baseUrl is that it should be an env variable in the first place - not a config variable. Even in the [docs](https://docs.cypress.io/guides/guides/environment-variables.html) it says:

> baseUrl can be set in your configuration file (cypress.json by default) - and then you can use an environment variable to override it.
> CYPRESS_baseUrl=https://staging.app.com cypress run

Cypress users need to override it very often hence it should be an env var - not a config var.

**Comment by markjaquith** [comment]
This has been annoying me for years.

This is probably a bit of a hack, but this works, for now. This is probably unsupported and a code time bomb, so caveat emptor:

`cypress/support/index.js`:
```js
// Allow baseUrl override.
import './baseUrl'

// Import commands.js using ES2015 syntax:
import './commands'
```

`cypress/support/baseUrl.js':
```js
// Allow a 'baseUrl' env var in cypress.env.json to override the Cypress baseUrl.
const baseUrl = Cypress.env('baseUrl');
baseUrl && Cypress.config('baseUrl', baseUrl);
```

`cypress.env.json`:
```js
{
	"baseUrl": "https://whatever.test",
}
```

**NOTE: You cannot set `baseUrl` in `cypress.json` if you do this (or rather, if you do, it will have to exist on the user's system, which defeats the whole purpose of letting them set it to something that exists on their system).

Because this is done in **support**, and outside spec files, `Cypress.config()` calls "stick", and stay set through all spec runs.

When you use the GUI to run tests, it's a little weird, because it first loads on `localhost:some-port` and then updates to your `https://whatever.test`. But it seems to work fine.

Ultimately, I wish Cypress would let you override config locally in a more elegant way. #5218 is probably the way forward.

**Comment by jennifer-shehane** [comment]
This issue will be closed to further comment as the exact issue here was resolved and tested in version 1.2.0.

If you're experiencing a bug similar to this in Cypress or want to request a change in behavior in how Cypress works today, please [open a new issue](https://github.com/cypress-io/cypress/issues/new) with a fully reproducible example that we can run. Closed issues are not the best place for discussion of wanted feature changes.
